src/main/java/core/basesyntax/ArrayList.java

```diff
@@ -1,48 +1,102 @@
 package core.basesyntax;
 
+import java.util.NoSuchElementException;
+
 public class ArrayList<T> implements List<T> {
+    private static final int DEFAULT_CAPACITY = 10;
+    private T[] elementData;
+    private int capacity;
+    private int size;
+
+    public ArrayList() {
+        elementData = (T[]) new Object[DEFAULT_CAPACITY];
+        capacity = DEFAULT_CAPACITY;
+        size = 0;
+    }
+
     @Override
     public void add(T value) {
-
+        add(value, size);
     }
 
     @Override
     public void add(T value, int index) {
-
+        if (index > size || index < 0) {
+            throw new ArrayListIndexOutOfBoundsException("The index passed is invalid");
+        }
+        if (size == capacity - 1) {
+            growIfNecessary();
+        }
+        if (index < size) {
+            System.arraycopy(elementData, index, elementData, index + 1, size - index);
+        }
+        elementData[index] = value;
+        size++;
     }
 
     @Override
     public void addAll(List<T> list) {
-
+        for (int i = 0; i < list.size(); i++) {
+            add(list.get(i));
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        verifyIndex(index);
+        return elementData[index];
     }
 
     @Override
     public void set(T value, int index) {
-
+        verifyIndex(index);
+        elementData[index] = value;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        verifyIndex(index);
+        size--;
+        T removedElement = elementData[index];
+        System.arraycopy(elementData, index + 1, elementData, index, size + 1 - index);
+        elementData[size] = null;
+        return removedElement;
     }
 
     @Override
     public T remove(T element) {
-        return null;
+        for (int i = 0; i < size; i++) {
+            if (element == null || elementData[i] != null
+                    && elementData[i].equals(element)) {
```

Comment:
The '==' operator should not be used to compare 'element' with 'null'. Instead, 'element == null' should be used to check for nullity explicitly. Also, the condition 'elementData[i] != null' is not needed because if 'element' is not null and 'elementData[i]' is null, the '.equals' method will not be called due to short-circuit evaluation.